src/app/config/confparse.c

```diff
@@ -388,13 +388,48 @@ static int
 config_validate_single(const config_format_t *fmt,
                        const void *oldval,
                        void *newval,
-                       char **msg_out)
+                       smartlist_t *errs_out)
 {
+  int res = CFG_OK;
+  if (fmt->pre_derive_fn) {
+    if (fmt->pre_derive_fn(newval) < 0) {
+      smartlist_add_strdup(errs_out, "Pre-validation derivation failed");
+      return CFG_INTERNAL_ERR;
+    }
+  }
+
   if (fmt->validate_fn) {
-    return fmt->validate_fn(oldval, newval, msg_out);
-  } else {
-    return 0;
+    if (fmt->validate_fn(newval, errs_out) < 0) {
+      res = CFG_BAD_VAL;
```

Comment:
Why do we return on a pre-derivation fail, but continue on a validate fail?
A comment would be useful here.

This change should be made shortly after merging this branch, because there should not be any more branches until after this branch is merged, and the outstanding bugs are fixed.

Reply:
Why doesn't validate_fn return a string error message?